File: DataEngineering/06_Data_Orchestration/projects/python/03_cron_job_scheduler/main.py
import time
import logging
from apscheduler.schedulers.background import BackgroundScheduler
from apscheduler.triggers.cron import CronTrigger

# Configure logging
logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s")
logger = logging.getLogger(__name__)

def run_data_pipeline(pipeline_name: str):
    logger.info(f"Starting pipeline: {pipeline_name}")
    # Simulate pipeline execution
    time.sleep(2)
    logger.info(f"Completed pipeline: {pipeline_name}")

if __name__ == "__main__":
    # Initialize the scheduler
    scheduler = BackgroundScheduler()

    # Schedule a task to run every minute
    scheduler.add_job(
        run_data_pipeline,
        trigger=CronTrigger(minute="*"), # Equivalent to cron '* * * * *'
        args=["Hourly_Sales_Aggregation"],
        id="sales_pipeline_job",
        replace_existing=True
    )

    logger.info("Starting scheduler. Press Ctrl+C to exit.")
    scheduler.start()

    try:
        # Keep the main thread alive to allow background jobs to run
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        logger.info("Shutting down scheduler...")
        scheduler.shutdown()

File: DataEngineering/06_Data_Orchestration/projects/golang/03_cron_job_scheduler/main.go
package main

import (
	"log"
	"os"
	"os/signal"
	"syscall"

	"github.com/robfig/cron/v3"
)

func runDataPipeline(pipelineName string) {
	log.Printf("Executing pipeline: %s", pipelineName)
	// In a real scenario, this would trigger actual data extraction and loading
}

func main() {
	log.Println("Starting cron job scheduler...")

	// Create a new cron instance
	// We use the standard cron syntax
	c := cron.New()

	// Add a job using standard cron syntax
	// "* * * * *" runs at the beginning of every minute
	_, err := c.AddFunc("* * * * *", func() {
		runDataPipeline("Hourly_Sales_Aggregation")
	})
	if err != nil {
		log.Fatalf("Failed to schedule job: %v", err)
	}

	// Start the cron scheduler in its own goroutine
	c.Start()

	log.Println("Scheduler is running. Press Ctrl+C to exit.")

	// Set up a channel to listen for interrupt signals to gracefully shutdown
	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)

	// Block until a signal is received
	<-sigChan

	log.Println("Shutting down scheduler...")
	c.Stop()
}
